### ISSUE TITLE
"fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree."

### ISSUE THREAD

**Issue by EdyQiu** [open-mmlab/mmdetection#5189]
The story is, I notice when I trained my model, there were warnings like "not git repository", but these messages seemed to have no actual effect on the training progress, so I usually just ignore it.

Yesterday, I suddenly want to remove this error message during training, and google told me to do "git init". I was so tired and just did that.

Then, this error message "**fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree**."  appears and the training process can not continue.
I totally had no clue, and just started from creating a new virtual environment in the "get start" document. 
After I downloaded everything, I trained my model and this error message still appear, though the training process seems fine.

So, I am worried about how could I remove this error message?
I would really appreciate any kind help!

**Comment by hhaAndroid** [comment]
Hi, Are you downloading the code directly? This prompt means that the current code is not managed by git？

**Comment by EdyQiu** [comment]
> Hi, Are you downloading the code directly? This prompt means that the current code is not managed by git？

Thanks for your reply! 
To be honest, I am not very familiar with git.
The error message firstly appeared when I trained my model after I ran the "git init" command.
I tried my best searching on the internet but could not solve it, or restore "git init".
So, I decided to do everything from scratch, and followed the "get start" document, 
specifically, create a new environment, install pytorch and mmcv, clone mmdetection(using git clone command, not unzipping) to my server and install things from "requirrement.txt".
Even Though I believe I "reset" everything, this error message still appears.

**Comment by EdyQiu** [comment]
More details are as follows, and I highlighted those important messeges

(mmdetection) qiuxinkuan@aa-ESC8000:~/dataset/VOCformat/generic_face_dataset/脸框有类别/JPEGImages$ CUDA_VISIBLE_DEVICES=2,3,5,6 PORT=29521 bash /home/qiuxinkuan/mmdetection/tools/dist_train.sh /home/qiuxinkuan/mmdetection/20210508_finetune_0419_classaware_face_detector/faster_rcnn_r101_fpn_1x_20210419_Facewithclass.py 1 --work-dir /home/qiuxinkuan/mmdetection/20210508_finetune_0419_classaware_face_detector
**fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'**
2021-05-21 04:04:33,078 - mmdet - INFO - Environment info:
------------------------------------------------------------
sys.platform: linux
Python: 3.7.10 (default, Feb 26 2021, 18:47:35) [GCC 7.3.0]
CUDA available: True
GPU 0,1,2,3: TITAN Xp
CUDA_HOME: :/home/qiuxinkuan/cuda-10.1/
GCC: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
PyTorch: 1.7.1
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.2-Product Build 20210312 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.6.0 (Git Hash 5ef631a030a6f73131c77892041042805a06064f)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_VULKAN_WRAPPER -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

TorchVision: 0.8.2
OpenCV: 4.5.2
MMCV: 1.3.4
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 10.1
MMDetection: 2.12.0+HEAD
------------------------------------------------------------

2021-05-21 04:04:34,803 - mmdet - INFO - Distributed training: True
2021-05-21 04:04:36,442 - mmdet - INFO - Config:
model = dict(
    type='FasterRCNN',
    pretrained=None,
    backbone=dict(
        type='ResNet',
        depth=101,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=-1,
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=True,
        style='pytorch'),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        num_outs=5),
    rpn_head=dict(
        type='RPNHead',
        in_channels=256,
        feat_channels=256,
        anchor_generator=dict(
            type='AnchorGenerator',
            scales=[8],
            ratios=[0.5, 1.0, 2.0],
            strides=[4, 8, 16, 32, 64]),
        bbox_coder=dict(
            type='DeltaXYWHBBoxCoder',
            target_means=[0.0, 0.0, 0.0, 0.0],
            target_stds=[1.0, 1.0, 1.0, 1.0]),
        loss_cls=dict(
            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0),
        loss_bbox=dict(type='L1Loss', loss_weight=1.0)),
    roi_head=dict(
        type='StandardRoIHead',
        bbox_roi_extractor=dict(
            type='SingleRoIExtractor',
            roi_layer=dict(type='RoIAlign', output_size=7, sampling_ratio=0),
            out_channels=256,
            featmap_strides=[4, 8, 16, 32]),
        bbox_head=dict(
            type='Shared2FCBBoxHead',
            in_channels=256,
            fc_out_channels=1024,
            roi_feat_size=7,
            num_classes=399,
            bbox_coder=dict(
                type='DeltaXYWHBBoxCoder',
                target_means=[0.0, 0.0, 0.0, 0.0],
                target_stds=[0.1, 0.1, 0.2, 0.2]),
            reg_class_agnostic=False,
            loss_cls=dict(
                type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0),
            loss_bbox=dict(type='L1Loss', loss_weight=1.0))),
    train_cfg=dict(
        rpn=dict(
            assigner=dict(
                type='MaxIoUAssigner',
                pos_iou_thr=0.7,
                neg_iou_thr=0.3,
                min_pos_iou=0.3,
                match_low_quality=True,
                ignore_iof_thr=-1),
            sampler=dict(
                type='RandomSampler',
                num=256,
                pos_fraction=0.5,
                neg_pos_ub=-1,
                add_gt_as_proposals=False),
            allowed_border=-1,
            pos_weight=-1,
            debug=False),
        rpn_proposal=dict(
            nms_pre=2000,
            max_per_img=1000,
            nms=dict(type='nms', iou_threshold=0.7),
            min_bbox_size=0),
        rcnn=dict(
            assigner=dict(
                type='MaxIoUAssigner',
                pos_iou_thr=0.5,
                neg_iou_thr=0.5,
                min_pos_iou=0.5,
                match_low_quality=False,
                ignore_iof_thr=-1),
            sampler=dict(
                type='RandomSampler',
                num=512,
                pos_fraction=0.25,
                neg_pos_ub=-1,
                add_gt_as_proposals=True),
            pos_weight=-1,
            debug=False)),
    test_cfg=dict(
        rpn=dict(
            nms_pre=1000,
            max_per_img=1000,
            nms=dict(type='nms', iou_threshold=0.7),
            min_bbox_size=0),
        rcnn=dict(
            score_thr=0.05,
            nms=dict(type='nms', iou_threshold=0.5),
            max_per_img=100)))
dataset_type = 'VOCDataset'
data_root = '/home/qiuxinkuan/dataset/VOCformat/generic_face_dataset'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='Resize', img_scale=(1000, 600), keep_ratio=True),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1000, 600),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='Pad', size_divisor=32),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ])
]
data = dict(
    samples_per_gpu=4,
    workers_per_gpu=4,
    train=dict(
        type='RepeatDataset',
        times=1,
        dataset=dict(
            type='VOCDataset',
            ann_file=[
                '/home/qiuxinkuan/dataset/VOCformat/generic_face_dataset/脸框有类别/ImageSets/Main/setting_finetune1.txt'
            ],
            img_prefix=[
                '/home/qiuxinkuan/dataset/VOCformat/generic_face_dataset/脸框有类别'
            ],
            pipeline=[
                dict(type='LoadImageFromFile'),
                dict(type='LoadAnnotations', with_bbox=True),
                dict(type='Resize', img_scale=(1000, 600), keep_ratio=True),
                dict(type='RandomFlip', flip_ratio=0.5),
                dict(
                    type='Normalize',
                    mean=[123.675, 116.28, 103.53],
                    std=[58.395, 57.12, 57.375],
                    to_rgb=True),
                dict(type='Pad', size_divisor=32),
                dict(type='DefaultFormatBundle'),
                dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
            ])),
    val=dict(
        type='VOCDataset',
        ann_file=[
            '/home/qiuxinkuan/dataset/VOCformat/generic_face_dataset/脸框有类别/ImageSets/Main/setting_test10.txt'
        ],
        img_prefix=
        '/home/qiuxinkuan/dataset/VOCformat/generic_face_dataset/脸框有类别',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=(1000, 600),
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='Pad', size_divisor=32),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ]),
    test=dict(
        type='VOCDataset',
        ann_file=[
            '/home/qiuxinkuan/dataset/VOCformat/generic_face_dataset/脸框有类别/ImageSets/Main/setting_test_base.txt',
            '/home/qiuxinkuan/dataset/VOCformat/generic_face_dataset/脸框有类别/ImageSets/Main/setting_test_novel.txt'
        ],
        img_prefix=
        '/home/qiuxinkuan/dataset/VOCformat/generic_face_dataset/脸框有类别',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=(1000, 600),
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='Pad', size_divisor=32),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ]))
evaluation = dict(interval=1, metric='mAP')
checkpoint_config = dict(interval=1)
log_config = dict(interval=50, hooks=[dict(type='TextLoggerHook')])
custom_hooks = [dict(type='NumClassCheckHook')]
dist_params = dict(backend='nccl')
log_level = 'INFO'
resume_from = None
workflow = [('train', 1)]
optimizer = dict(type='SGD', lr=0.001, momentum=0.9, weight_decay=0.0001)
optimizer_config = dict(grad_clip=None)
lr_config = dict(policy='step', step=[3])
runner = dict(type='EpochBasedRunner', max_epochs=20)
work_dir = '/home/qiuxinkuan/mmdetection/20210508_finetune_0419_classaware_face_detector'
load_from = '/home/qiuxinkuan/mmdetection/20210508_finetune_0419_classaware_face_detector/ptm.pth'
gpu_ids = range(0, 1)

**/home/qiuxinkuan/anaconda3/envs/mmdetection/lib/python3.7/site-packages/mmcv/cnn/utils/weight_init.py:100: UserWarning: init_cfg without layer key, if you do not define override key either, this init_cfg will do nothing
  'init_cfg without layer key, if you do not define override'
fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'**
2021-05-21 04:04:38,836 - mmdet - INFO - load checkpoint from /home/qiuxinkuan/mmdetection/20210508_finetune_0419_classaware_face_detector/ptm.pth
2021-05-21 04:04:38,837 - mmdet - INFO - Use load_from_local loader
2021-05-21 04:04:39,417 - mmdet - WARNING - The model and loaded state dict do not match exactly

size mismatch for roi_head.bbox_head.fc_cls.weight: copying a param with shape torch.Size([320, 1024]) from checkpoint, the shape in current model is torch.Size([400, 1024]).
size mismatch for roi_head.bbox_head.fc_cls.bias: copying a param with shape torch.Size([320]) from checkpoint, the shape in current model is torch.Size([400]).
size mismatch for roi_head.bbox_head.fc_reg.weight: copying a param with shape torch.Size([1276, 1024]) from checkpoint, the shape in current model is torch.Size([1596, 1024]).
size mismatch for roi_head.bbox_head.fc_reg.bias: copying a param with shape torch.Size([1276]) from checkpoint, the shape in current model is torch.Size([1596]).
2021-05-21 04:04:39,434 - mmdet - INFO - Start running, host: qiuxinkuan@aa-ESC8000, work_dir: /home/qiuxinkuan/mmdetection/20210508_finetune_0419_classaware_face_detector
2021-05-21 04:04:39,434 - mmdet - INFO - workflow: [('train', 1)], max: 20 epochs
2021-05-21 04:04:49,152 - mmcv - INFO - Reducer buckets have been rebuilt in this iteration.
2021-05-21 04:05:22,905 - mmdet - INFO - Epoch [1][50/101]      lr: 1.000e-03, eta: 0:28:32, t

**Comment by EdyQiu** [comment]
Also, I noticed that some mmdetection files in VS code now become "grey" ,
I guess it may related to this issue.

![image](https://user-images.githubusercontent.com/23387293/119111290-80c03a00-ba55-11eb-9199-daf01e89d76a.png)

**Comment by EdyQiu** [comment]
The "init_cfg " issue is due to the update, though I have not figured out how to solve that.
as for the "HEAD" issue, I just discard my old files, created a new folder and start everything from scratch. e.g. install cuda
and currently this error msg disappeared
